build/Dockerfile

```diff
@@ -72,88 +64,91 @@ ENV CUDA_HOME="/usr/local/cuda" \
     PATH="/usr/local/nvidia/bin:${CUDA_HOME}/bin:${PATH}" \
     LD_LIBRARY_PATH="/usr/local/nvidia/lib:/usr/local/nvidia/lib64:$CUDA_HOME/lib64:$CUDA_HOME/extras/CUPTI/lib64:${LD_LIBRARY_PATH}"
 
+## CUDA Development ############################################################
+FROM cuda-base as cuda-devel
 
-ENV NV_NVTX_VERSION=11.8.86-1 \
-    NV_LIBNPP_VERSION=11.8.0.86-1 \
-    NV_LIBCUBLAS_VERSION=11.11.3.6-1 \
-    NV_LIBNCCL_PACKAGE_VERSION=2.15.5-1+cuda11.8
+# Ref: https://developer.nvidia.com/nccl/nccl-legacy-downloads
+ENV NV_CUDA_CUDART_DEV_VERSION=12.1.55-1 \
+    NV_NVML_DEV_VERSION=12.1.55-1 \
+    NV_LIBCUBLAS_DEV_VERSION=12.1.0.26-1 \
+    NV_LIBNPP_DEV_VERSION=12.0.2.50-1 \
+    NV_LIBNCCL_DEV_PACKAGE_VERSION=2.18.3-1+cuda12.1
 
 RUN dnf config-manager \
        --add-repo https://developer.download.nvidia.com/compute/cuda/repos/rhel9/x86_64/cuda-rhel9.repo \
     && dnf install -y \
-        cuda-libraries-11-8-${NV_CUDA_LIB_VERSION} \
-        cuda-nvtx-11-8-${NV_NVTX_VERSION} \
-        libnpp-11-8-${NV_LIBNPP_VERSION} \
-        libcublas-11-8-${NV_LIBCUBLAS_VERSION} \
-        libnccl-${NV_LIBNCCL_PACKAGE_VERSION} \
-    && dnf clean all
-
-ENV NV_CUDA_CUDART_DEV_VERSION=11.8.89-1 \
-    NV_NVML_DEV_VERSION=11.8.86-1 \
-    NV_LIBCUBLAS_DEV_VERSION=11.11.3.6-1 \
-    NV_LIBNPP_DEV_VERSION=11.8.0.86-1 \
-    NV_LIBNCCL_DEV_PACKAGE_VERSION=2.15.5-1+cuda11.8
-
-RUN dnf config-manager \
-       --add-repo https://developer.download.nvidia.com/compute/cuda/repos/rhel9/x86_64/cuda-rhel9.repo \
-    && dnf install -y \
-        cuda-command-line-tools-11-8-${NV_CUDA_LIB_VERSION} \
-        cuda-libraries-devel-11-8-${NV_CUDA_LIB_VERSION} \
-        cuda-minimal-build-11-8-${NV_CUDA_LIB_VERSION} \
-        cuda-cudart-devel-11-8-${NV_CUDA_CUDART_DEV_VERSION} \
-        cuda-nvml-devel-11-8-${NV_NVML_DEV_VERSION} \
-        libcublas-devel-11-8-${NV_LIBCUBLAS_DEV_VERSION} \
-        libnpp-devel-11-8-${NV_LIBNPP_DEV_VERSION} \
+        cuda-command-line-tools-12-1-${NV_CUDA_LIB_VERSION} \
+        cuda-libraries-devel-12-1-${NV_CUDA_LIB_VERSION} \
+        cuda-minimal-build-12-1-${NV_CUDA_LIB_VERSION} \
+        cuda-cudart-devel-12-1-${NV_CUDA_CUDART_DEV_VERSION} \
+        cuda-nvml-devel-12-1-${NV_NVML_DEV_VERSION} \
+        libcublas-devel-12-1-${NV_LIBCUBLAS_DEV_VERSION} \
+        libnpp-devel-12-1-${NV_LIBNPP_DEV_VERSION} \
         libnccl-devel-${NV_LIBNCCL_DEV_PACKAGE_VERSION} \
     && dnf clean all
 
 ENV LIBRARY_PATH="$CUDA_HOME/lib64/stubs"
 
-RUN dnf install -y python3.11 git \
-    && ln -s /usr/bin/python3.11 /bin/python \
-    && python -m ensurepip --upgrade \
-    && dnf clean all
+FROM cuda-devel as python-installations
 
-# Removes the example private key to avoid high severity vulnerability warning
-RUN rm -f /usr/share/doc/perl-Net-SSLeay/examples/server_key.pem
+ARG WHEEL_VERSION
+ARG USER
 
+USER ${USER}
```

Comment:
Interesting that python installs are all happening as USER instead of as root. Previously, we've built everything as root and only at the final stage would we run the image as root. Would this pose any issues if someone runs the image as a different user? I wouldn't think so but want to check

Reply:
The tuning user is in the root group (-g 0), see [here](https://github.com/tharapalanivel/fms-hf-tuning/blob/e7a47f8e4598690ee663e7406baa7344e6900413/build/Dockerfile). Any user in the root group can access python packages installed by this user.
Unless someone specify non-zero `runAsGroup` in [scc](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#set-the-security-context-for-a-pod) or openshift runs pod as random user from root group. For example, when I log in the runing pod I ses:
```
bash-5.1$ whoami
1000690000
bash-5.1$ cat /etc/passwd
root:x:0:0:root:/root:/bin/bash
...
tuning:x:1000:0::/home/tuning:/bin/bash
1000690000:x:1000690000:0:1000690000 user:/:/sbin/nologin
```

Reply:
So as long as the user that is running the image is in the root group we should be good, right? In which when the image is run on Openshift, it defaults to using the tuning user correct? Or does it default to a random user?

Reply:
> as long as the user that is running the image is in the root group we should be good, right?

correct as long as user is in root group we are good.

> In which when the image is run on Openshift, it defaults to using the tuning user correct? Or does it default to a random user?

User should be random But In our OpenShift, the user default to the id in the namespace CR. For example 
```bash
oc get ns fmaas-dev -oyaml
apiVersion: v1
kind: Namespace
metadata:
  name: fmaas-dev
  annotations:
    ...
    openshift.io/sa.scc.supplemental-groups: 1000690000/10000
    openshift.io/sa.scc.uid-range: 1000690000/10000
    ...
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
```